lib/Map/Vector/ProtomapsGeojsonSource.ts

```diff
@@ -82,73 +82,133 @@ export class ProtomapsGeojsonSource implements TileSource {
-              props: { ...(f.tags ?? {}) },
-              bbox,
-              geomType,
-              geom: transformedGeom,
-              numVertices
-            };
-
-            return feature;
-          })
-          .filter((f) => f !== null) as ProtomapsFeature[]
+        geojsonVtTileToProtomapsFeatures(tile.features, tileSize)
       );
     }
 
     return result;
   }
 }
+
+export const geomTypeMap = (
+  type: string | null | undefined
+): GeomType | null => {
+  switch (type) {
+    case "Point":
+    case "MultiPoint":
+      return GeomType.Point;
+    case "LineString":
+    case "MultiLineString":
+      return GeomType.Line;
+    case "Polygon":
+    case "MultiPolygon":
+      return GeomType.Polygon;
+    default:
+      return null;
+  }
+};
+
+export function geojsonVtTileToProtomapsFeatures(
+  features: geojsonvt.Feature[],
+  tileSize: number
+): ProtomapsFeature[] {
+  const scale = tileSize / GEOJSON_VT_EXTENT;
+
+  return features
+    .map((f) => {
+      let transformedGeom: Point[][] = [];
+      let numVertices = 0;
+
+      // Calculate bbox
+      const bbox: Bbox = {
+        minX: Infinity,
+        minY: Infinity,
+        maxX: -Infinity,
+        maxY: -Infinity
+      };
+
+      // Multi-polygon
+      if (Array.isArray(f.geometry[0][0])) {
+        // Note: the type is incorrect here
+        const geom = f.geometry as unknown as [number, number][][];
+        transformedGeom = geom.map((g1) =>
+          g1.map((g2) => {
+            const x = g2[0] * scale;
+            const y = g2[1] * scale;
+
+            if (bbox.minX > x) {
+              bbox.minX = x;
+            }
+
+            if (bbox.maxX < x) {
+              bbox.maxX = x;
+            }
+
+            if (bbox.minY > y) {
+              bbox.minY = y;
+            }
+
+            if (bbox.maxY < y) {
+              bbox.maxY = y;
+            }
+            return new Point(x, y);
+          })
+        );
+        numVertices = transformedGeom.reduce<number>(
+          (count, current) => count + current.length,
+          0
+        );
+      }
+      // Other feature types
+      else {
+        const geom = f.geometry as [number, number][];
+        transformedGeom = [
+          geom.map((g1) => {
+            const x = g1[0] * scale;
+            const y = g1[1] * scale;
+
+            if (bbox.minX > x) {
+              bbox.minX = x;
+            }
+
+            if (bbox.maxX < x) {
+              bbox.maxX = x;
+            }
+
+            if (bbox.minY > y) {
+              bbox.minY = y;
+            }
+
+            if (bbox.maxY < y) {
+              bbox.maxY = y;
+            }
+            return new Point(x, y);
+          })
+        ];
+        numVertices = transformedGeom.length;
+      }
+
+      if (f.type === 0) return null;
+
+      const geomType = {
+        [1]: GeomType.Point,
+        [2]: GeomType.Line,
+        [3]: GeomType.Polygon
+      }[f.type];
```

Comment:
Is this the bug fix?

I don't have a strong opinion on whether a PR marked as bug fix should contain minor refactors or not, but PRs are easier to review if the bug fix is in a separate commit from the commit that moves a big chunk of code around.

Reply:
I think it is fine because this change is cherry picked from [a bigger PR](https://github.com/TerriaJS/terriajs/pull/7370) for a bug fix release. More refactoring changes will follow in the parent PR.